customlists/customlist_export.py

```diff
@@ -0,0 +1,484 @@
+    def _parse_arguments(args: List[str]) -> argparse.Namespace:
+        parser: argparse.ArgumentParser = argparse.ArgumentParser(
+            description="Fetch one or more custom lists."
+        )
+        parser.add_argument(
+            "--schema-file",
+            help="The path to customlists.schema.json",
+            required=False,
+            default="customlists/customlists.schema.json",
+        )
+        parser.add_argument("--server", help="The address of the CM", required=True)
+        parser.add_argument("--username", help="The CM admin username", required=True)
+        parser.add_argument("--password", help="The CM admin password", required=True)
+        parser.add_argument("--output", help="The output file", required=True)
+        parser.add_argument(
+            "--list", help="Only export the named list (may be repeated)", nargs="+"
+        )
+        parser.add_argument(
+            "--verbose",
+            "-v",
+            action="count",
+            default=0,
+            help="Increase verbosity (can be specified multiple times to export multiple lists)",
+        )
+        return parser.parse_args(args)
+
+    def _make_custom_list(
+        self, raw_list: dict
+    ) -> Union[CustomList, ProblematicCustomList]:
+        id: int = raw_list["id"]
+        name: str = raw_list["name"]
+
+        # The /admin/custom_list/ URL will yield an OPDS feed of the list contents.
+        server_list_endpoint: str = f"{self._server_base}/admin/custom_list/{id}"
+        response = self._session.get(server_list_endpoint)
+        if response.status_code >= 400:
+            return ProblematicCustomList(
+                id=id,
+                name=name,
+                error=f"Failed to retrieve custom list {id}: {response.status_code} {response.reason}",
+            )
+
+        feed = feedparser.parse(url_file_stream_or_string=response.content)
+
+        # Extract the "self" link in order to determine the library identifier.
+        library_short_id: Optional[str] = None
+        for link in feed.feed.links:
+            if link.rel == "self":
+                match = re.search(f"^(.*)/(.*)/admin/custom_list/{id}(.*)", link.href)
+                if match is not None:
+                    library_short_id = match.group(2)
+                    break
+
+        # If we couldn't find a self link, then we don't know which library has this list.
+        if not library_short_id:
+            return ProblematicCustomList(
+                id, name, "Unable to locate a 'self' link in the custom list feed"
+            )
+
+        custom_list = CustomList(id=id, name=name, library_id=library_short_id)
+
+        # Now, for each book, extract the book identifier and identifier type.
+        for entry in feed.entries:
+            added = False
+            for link in entry.links:
+                if link.rel == "alternate":
+                    match = re.search("^(.*)/works/([^/]+)/(.*)$", link.href)
```

Comment:
To my knowledge, these are consistent, because they're generated by the CM internally and don't appear to be part of data that came from some third party.